pkg/stenciltest/stenciltest.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2022 Outreach Corporation. All Rights Reserved.
+
+// Description: This file implements the stenciltest framework
+// for testing templates generated by stencil.
+
+// Package stenciltest contains code for testing templates
+package stenciltest
+
+import (
+	"context"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/bradleyjkemp/cupaloy"
+	"github.com/getoutreach/stencil/internal/codegen"
+	"github.com/getoutreach/stencil/internal/modules"
+	"github.com/getoutreach/stencil/internal/modules/modulestest"
+	"github.com/getoutreach/stencil/pkg/configuration"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v3"
+)
+
+// Template is a template that is being tested by the stenciltest framework.
+type Template struct {
+	// path is the path to the template.
+	path string
+
+	// aditionalTemplates is a list of additional templates to add to the renderer,
+	// but not to snapshot.
+	additionalTemplates []string
+
+	// m is the template repository manifest for this test
+	m *configuration.TemplateRepositoryManifest
+
+	// t is a testing object.
+	t *testing.T
+
+	// args are the arguments to the template.
+	args map[string]interface{}
+}
+
+// New creates a new test for a given template.
+func New(t *testing.T, templatePath string, additionalTemplates ...string) *Template {
+	// GOMOD: <module path>/go.mod
+	b, err := exec.Command("go", "env", "GOMOD").Output()
+	if err != nil {
+		t.Errorf("failed to determine path to manifest: %v", err)
+	}
+	basepath := strings.TrimSuffix(strings.TrimSpace(string(b)), "/go.mod")
+
+	b, err = os.ReadFile(filepath.Join(basepath, "manifest.yaml"))
+	if err != nil {
+		t.Error(err)
+	}
+
+	var m configuration.TemplateRepositoryManifest
+	if err := yaml.Unmarshal(b, &m); err != nil {
+		t.Error(err)
+	}
+
+	return &Template{
+		t:                   t,
+		m:                   &m,
+		path:                templatePath,
+		additionalTemplates: additionalTemplates,
+	}
+}
+
+// Args sets the arguments to the template.
+func (t *Template) Args(args map[string]interface{}) *Template {
+	t.args = args
+	return t
+}
+
+// Run runs the test.
+func (t *Template) Run(save bool) {
+	t.t.Run(t.path, func(got *testing.T) {
+		m, err := modulestest.NewModuleFromTemplates(t.m.Arguments, append([]string{t.path}, t.additionalTemplates...)...)
+		if err != nil {
+			got.Errorf("failed to create module from template %q", t.path)
+		}
+
+		mf := &configuration.ServiceManifest{Name: "testing", Arguments: t.args,
+			Modules: []*configuration.TemplateRepository{{Name: m.Name}}}
+		st := codegen.NewStencil(mf, []*modules.Module{m})
+
+		tpls, err := st.Render(context.Background(), logrus.New())
+		if err != nil {
```

Comment:
Will there be a case where the template owner will want to validate that an error is received?

For Temporal as an example we previously had as part of the template validation of the versions and failed if it was an unsupported one. 

This is probably an edge-case the scenario I'm pointing out but in a roundabout way I'm wondering if we'd want to validate an error occurs in certain templates.

Reply:
That's a fair ask! That makes sense, I'll add that in.

Reply:
Added `ErrorContains` for this usecase!